core/trino-main/src/test/java/io/trino/sql/planner/AbstractPredicatePushdownTest.java

```diff
@@ -66,14 +67,15 @@ public void testNonStraddlingJoinExpression()
         assertPlan(
                 "SELECT * FROM orders JOIN lineitem ON orders.orderkey = lineitem.orderkey AND cast(lineitem.linenumber AS varchar) = '2'",
                 anyTree(
-                        join(INNER, ImmutableList.of(equiJoinClause("ORDERS_OK", "LINEITEM_OK")),
+                        join(new JoinParamUtil.JoinParamBuilder(INNER,
```

Comment:
Hi @lukasz-stec The only way I am thinking of having a `JoinMatcherBuilder` which will have static method with params `joinType` & `equiCriteria` to create object of `JoinMatcher`, and setter for other attributes. 
Like below,

`join(JoinMatcherBuilder.buildMatcher(joinType, equiCriteria).filter(filter), leftPlanMatchPattern, rightPlanMatchPattern)`

But that will require to change `JoinMatcher` from immutable to mutable.
Otherwise, avoiding method overloading to use builder pattern will bring code verbosity. Please guide what can be done?

Reply:
to remove `new JoinParamUtil.JoinParamBuilder` you could have a single class, `JoinMatcherBuilder`

```
class JoinMatcherBuilder
{
    public static JoinMatcherBuilder join(
      JoinNode.Type joinType,
      List<ExpectedValueProvider<JoinNode.EquiJoinClause>> expectedEquiCriteria,
      PlanMatchPattern left,
      PlanMatchPattern right) {..}

  ...

    PlanMatchPattern build() {...}
}
```

That still requires `build` for every current `PlanMatchPattern.join` method call.
One way to keep it out of tests is to add duplicated method to every method in `PlanMatchPattern` that takes PlanMatchPattern as param to instead take `PlanMatchPatternBuilder` (new interface) but that seems too much just for joins.
@sopel39 you're the original issue author. Did you have some ideas on what this should look like when you created the ticket?